tidalsim/bb/common.py

```diff
@@ -1,9 +1,72 @@
+import bisect
 from dataclasses import dataclass
-from intervaltree import IntervalTree, Interval
+from typing import List, Tuple, Optional, cast
 
 @dataclass
 class BasicBlocks:
-    pc_to_bb_id: IntervalTree
+    markers: List[Tuple[int, Optional[int]]]
+
+    def pc_to_bb_id(self, pc: int) -> Optional[int]:
+        bisect_index = bisect.bisect(self.markers, pc, key=lambda x: x[0]) - 1
+
+        if bisect_index == len(self.markers):
+            return None
+
+        _, basic_block = self.markers[bisect_index]
+        return basic_block
+
+    def __len__(self):
+        # Counts the number of markers that map to the start of a basic block
+        return len([bb_idx for _, bb_idx in self.markers if bb_idx is not None])
+
+# Tuple of [left, right), where the left is inclusive and the right is not.
+Interval = Tuple[int, int]
+
+# Tuple of the location of the event (which starts a new basic block), and up to what non-inclusive location
+# we consider it a valid code path. This is useful since if the location lies outside of the valid codepaths
+# by the blocks that prcede it, we can mark it as being empty space rather than being a BasicBlock.
+Event = Tuple[int, int]
+
+# Tuple of the start of a basic block and the id it maps to. If the basic block id is None, then it is not
+# really a basic block but empty space.
+Marker = Tuple[int, int | None]
+
+def intervals_to_events(intervals: List[Interval]) -> List[Event]:
+    events: List[Tuple[int, int]] = []
+    for start, end in intervals:
+        events += [(start, end), (end, 0)]
+    return events
+
+def events_to_markers(events: List[Event]) -> List[Marker]:
+    left = 0
+    right = 0
+    idx = 0
+    markers = []
+
+    # We sort events from left-to-right, but only really care about the highest value at each location
+    for pc, valid in sorted(events, key=lambda tup: (tup[0], -tup[1])):
```

Comment:
Something is broken here when using `--elf`.

```
  File "/scratch/vighneshiyer/chipyard/tools/tidalsim/tidalsim/bb/common.py", line 47, in <lambda>
    for pc, valid in sorted(events, key=lambda tup: (tup[0], -tup[1])):
TypeError: bad operand type for unary -: 'NoneType'
```

Could you test this by cloning Chipyard (https://github.com/ucb-bar/chipyard) (follow instructions here: https://github.com/euphoric-hardware/tidalsim#setup-in-chipyard) and building the embench benchmarks `cd software/embench && ./build.sh`, and then invoke tidalsim like `tidalsim --elf --binary ../../software/embench/build/aha-mont64 --interval-length 1000 --clusters 18 --simulator ../../sims/vcs/simv-inject-chipyard.harness-RocketConfig --chipyard-root ../../. --dest-dir runs`.